TOT-96 : [REFACTOR] 계획 작성 로직 리팩토링

honee8583 opened the pull request:
## PR전 체크리스트
> 1. PR 및 Commit title을 형식에 맞게 작성했나요(e.g. TOT-123 : [TYPE]) O
> 2. Reviewers가 명확하게 지정됐나요? O
> 3. Assignees가 명확하게 지정됐나요? O


## 📝 작업 내용
- 테스트 코드 리팩토링
- PlanServiceTest 테스트 코드 `@SpringBootTest` -> `@ExtendWith(MockitoExtension.class)` 로 변경.
- PlanServiceTest의 각 가짜객체들 `@MockBean` -> `@Mock`으로 변경.
- Plan, Schedule, SchedulePlace 요청 Dto클래스를 PlanRequestDto클래스로 통합후 내부 클래스로 구현.
- `@Nested`로 계획 작성 테스트 코드와 계획 상세 조회 테스트 코드를 분리.

### 스크린샷

## 💬 리뷰 요구사항
- 계획에 대한 요청 Dto를 하나로 통합하는게 괜찮은지? 
(계획작성, 계획수정, 계획상세정보에 대한 Dto들이 개별로 존재하게 되면 Dto클래스들이 너무 많아질거 같아 통합했습니다.) 
- 만약에 괜찮다면 계획작성Dto, 계획수정Dto를 나눌지 혹은 PlanRequestDto 클래스안에 계획작성Dto들과 계획수정Dto들을 모두 포함시키는게 좋을지? 
(작성과 수정에 대한 Dto클래스를 나눈다면 PlanRequestDto클래스의 클래스명을 수정해서 반영하겠습니다)
- 참고자료에 예시코드를 올려놓았는데 테스트 코드에서 더미데이터를 만들때 사용할 protected 메소드를 모아둔 클래스를 따로 만들어두고 테스트 클래스에서 상속시키면 한번더 분리할 수 있다는 것을 알았는데 어떻게 생각하시는지 궁금합니다!


## 참고자료
```java
public class DummyObject {
    protected Plan createPlan(Long id, String title) {
        return Plan.builder()
                .id(id)
                .title(title)
                .build();
    }
}


public class PlanServiceTest extends DummyObect {
    ...

    Plan plan = createPlan();
}
```

honee8583 committed:
TOT-96 : [TEST] 계획 작성 테스트 코드 리팩토링

- 계획작성, 계획상세정보조회 테스트 코드를 클래스 단위로 분리.

honee8583 committed:
TOT-96 : [TEST] 계획 작성,조회 테스트 코드 리팩토링

honee8583 committed:
TOT-96 : [TEST] 계획 컨트롤러 테스트 클래스에 ObjectMapper 주입

honee8583 committed:
TOT-96 : [REFACTOR] 계획 작성 요청 Dto 클래스들을 하나의 Dto로 수정.

- PlanRequestDto 클래스 내부에 PlanCreateRequestDto, ScheduleCreateRequestDto, SchedulePlaceCreateRequestDto 내부 클래스들을 추가.
- 기존 Dto들 삭제.

honee8583 committed:
TOT-96 : [REFACTOR] Dto클래스 변경에 따른 PlanService import문 수정

honee8583 committed:
TOT-96 : [REFACTOR] Dto클래스 변경에 따른 PlanController import문 수정

honee8583 committed:
TOT-96 : [REFACTOR] Dto클래스 변경에 따른 테스트 코드 import문 수정

snowcrab382 commented on backend/src/test/java/com/triportreat/backend/plan/controller/PlanControllerTest.java:
```diff
@@ -42,87 +42,95 @@ class PlanControllerTest {
-                .andExpect(jsonPath("$.message", equalTo(VALIDATION_FAILED.getMessage())))
-                .andExpect(jsonPath("$.result", equalTo(false)));
+        @Test()
+        @DisplayName("실패 - 유효성 검증 실패")
```
이렇게 테스트 이름을 깔끔하게 작성하니 가독성이 훨씬 뛰어난 것 같습니다! 다른 모든 테스트도 이와 같이 네이밍을 통일하면 좋을 것 같네요

honee8583 replied:
넵! 제가 구현했던 다른 기능들의 로직들도 한번씩 리팩토링할 예정이라 그때 통일시킬 예정입니다!

snowcrab382 commented on backend/src/test/java/com/triportreat/backend/plan/service/PlanServiceTest.java:
```diff
@@ -36,108 +37,112 @@
-    PlanService planService;
+    @InjectMocks
+    PlanServiceImpl planService;
 
-    @MockBean
+    @Mock
     PlanRepository planRepository;
 
-    @MockBean
+    @Mock
     UserRepository userRepository;
 
-    @MockBean
+    @Mock
     ScheduleRepository scheduleRepository;
 
-    @MockBean
+    @Mock
     PlaceRepository placeRepository;
 
-    @MockBean
+    @Mock
     SchedulePlaceRepository schedulePlaceRepository;
```
제가 찾아본 내용으로는 "테스트시 SpringBootTest가 사용된 클래스를 테스트할 때마다 애플리케이션 실행 전 필요한 의존성 주입이 매번 일어나는데 ExtendWith을 사용하면 해당 테스트에 필요한 객체들만 Mock으로 주입받아 테스트 수행 속도를 개선할 수 있다"인데 이 목적으로 적용하신게 맞을까요?? 맞다면 얼마나 개선되는지도 궁금하네요!

honee8583 replied:
네 맞습니다! 아이러니하게도 인텔리제이에서 보여주는 테스트가 걸린 시간은 더 오래걸렸다고 나오는데 제가 체감한 걸린시간은 훨씬 짧습니다;;

snowcrab382 commented:
고생하셨습니다! 테스트코드 관련 개선점이 많아서 도움이 많이 되었습니다.

아래는 리뷰 요구사항에 대한 답변입니다.
> 계획에 대한 요청 Dto를 하나로 통합하는게 괜찮은지?

저도 계획 수정관련 dto를 생성할 때 통합하는게 좋을 것 같다는 생각을 했습니다. 동의합니다!

> 만약에 괜찮다면 계획작성Dto, 계획수정Dto를 나눌지 혹은 PlanRequestDto 클래스안에 계획작성Dto들과 계획수정Dto들을 모두 포함시키는게 좋을지?

여기서 궁금한 게 있습니다. 혹시 계획작성과 수정Dto를 공통으로 쓸 수는 없을까요?? 계획 수정에 필요한 필드들이 모두 계획 작성 관련 Dto에도 동일하게 포함되어있어서 작성과 수정 Dto를 구분할 필요는 없을 것 같다는 생각입니다.

> 참고자료에 예시코드를 올려놓았는데 테스트 코드에서 더미데이터를 만들때 사용할 protected 메소드를 모아둔 클래스를 따로 만들어두고 테스트 클래스에서 상속시키면 한번더 분리할 수 있다는 것을 알았는데 어떻게 생각하시는지 궁금합니다!

이렇게 클래스를 따로 둔다면 테스트용 데이터를 만들 때 중복 로직이 감소하는 효과가 있어 좋을 것 같은데 이를 사용할 때 발생하는 문제점은 특별히 없을까요?? 단점이 없다면 매우 좋은 방법인 것 같습니다!

honee8583 commented:
- 계획작성Dto와 계획수정Dto를 하나로 사용하게 되면 `planId`, `scheduleId`, `SchedulePlaceId` 필드만 추가하면 되지만 해당 값들에 대한 validation 체크는 못하게 될것 같네요. 이 부분에 대해서 감안해도 괜찮다고 생각하시면 공통으로 사용해도 괜찮을 것 같습니다!
- 테스트 더미데이터용 클래스에 대한 단점은 아직까진 발견하지 못한것 같습니다. 사용하게 되면 눈에 띄는 단점을 발견했을 때 다시 돌려놓는 것도 좋을 것 같습니다!

snowcrab382 commented:
> 계획작성Dto와 계획수정Dto를 하나로 사용하게 되면 `planId`, `scheduleId`, `SchedulePlaceId` 필드만 추가하면 되지만 해당 값들에 대한 validation 체크는 못하게 될것 같네요. 이 부분에 대해서 감안해도 괜찮다고 생각하시면 공통으로 사용해도 괜찮을 것 같습니다!

그렇다면 우선 분리해서 작성해주시고 제가 계획수정 개발 진행하면서 통합해도 괜찮다고 생각이 들면 그때 통합하는 것으로 하는 게 좋을 것 같습니다!

MisaSohee commented:
제가 일정이 쭉 있었어서 너무 늦게 봤네요ㅜㅜ! Dto부분은 동훈님과 다훈님이 나누신 얘기에 동의합니다. @ExtendWith 사용과 @Nested 등등 테스트 코드 또한 잘 작성해주셔서 참고해서 진행하겠습니다 감사해요 고생하셨습니다!

honee8583 committed:
TOT-96 : [REFACTOR] 테스트용 Mock엔티티 및 생성Dto를 위한 DummyObject 클래스 생성 및 적용

- 테스트용 User, Plan, Schedule, SchedulePlace 엔티티를 생성하기 위한 메소드들 추가.
- 여러가지 상황들을 위해 매개변수로 필드값을 삽입할 수 있도록 생성.
- 중요하지 않은 필드이거나 변할 필요가 없는 필드들은 고정값으로 세팅.
  Plan: title 고정, name,nickname,email필드는 name필드 기반으로 설정
  SchedulePlace: expense 고정
  PlanRequestDto: title 고정
  SchedulePlaceRequestDto: memo,expense 고정

snowcrab382 reviewed (approved):
고생하셨습니다!

MisaSohee commented:
테스트코드가 훨씬 깔끔해졌네요..! 참고하겠습니다 고생하셨습니다
